src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -427,6 +433,8 @@ public string Substring(int offset, int length)
         /// <paramref name="offset"/> or <paramref name="length"/> is less than zero, or <paramref name="offset"/> + <paramref name="length"/> is
         /// greater than <see cref="Length"/>.
         /// </exception>
+        [MemberNotNull(nameof(Buffer))]
+        [MemberNotNull(nameof(Value))]
```

Comment:
`Subsegment(...)` is an odder case because the `StringSegment` will have non-`null` `Buffer` and `Value` properties but only when the method doesn't `throw`. Again however, this `StringSegment` isn't changed.

Reply:
From the [doc](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.membernotnullattribute?view=net-5.0):
> Specifies that the method or property will ensure that the listed field and property members have values that aren't null.

Method doesn't need to modify members, just ensure that they are initialized.

If it throws members are not considered as initialized (contain non-null values) as I showed here: https://github.com/dotnet/runtime/pull/58308#discussion_r697905372